Add cvmfsexec (SOFTWARE-4549)

matyasselmeci opened the pull request:
- Use cvmfsexec to mount CVMFS repos if needed
- Don't run crond (doesn't work as non-root)
- Create log files before tailing them to solve "<FILE> has been replaced witha remote file" error from tail

brianhlin reviewed (commented):
Why add this as `entrypoint.sh` instead of adding it a script to `/etc/osg/image-init.d`? @DrDaveD could you take a look at the `cvmfsexec` bits?

edquist commented:
Zing.

matyasselmeci commented:
Because I want cvmfs to be mounted regardless of the command being run.

matyasselmeci commented:
In this case, entrypoint.sh is used as the docker ENTRYPOINT, not the CMD; cvmfsexec is actually a wrapper around running some other command. I don't think it will work if I put it into /etc/osg/image-init.d since it does not affect the currently running process.

brianhlin commented:
> In this case, entrypoint.sh is used as the docker ENTRYPOINT, not the CMD; cvmfsexec is actually a wrapper around running some other command. I don't think it will work if I put it into /etc/osg/image-init.d since it does not affect the currently running process.

Oh, so the idea is that users of this container can pass runtime args to `cvmfsexec`? Could you add some comments to the `Dockerfile` to explain why we're doing it this way?

matyasselmeci commented:
I'll update the readme since it looks like it is confusing.

You're not passing runtime arguments to cvmfsexec (except via the new environment variables CVMFSEXEC_DIST and CVMFSEXEC_REPOS. Instead, you are passing runtime arguments to whatever command you're running.

For example,
`docker run ... opensciencegrid/osgvo-docker-pilot`
will run supervisord with CVMFS mounted.
`docker run ... opensciencegrid/osgvo-docker-pilot bash`
will run bash with CVMFS mounted.

on the other hand,
`docker run ... --entrypoint /bin/bash openscienceegrid/osgvo-docker-pilot`
will run bash without going through entrypoint.sh, so CVMFS won't be mounted.

brianhlin reviewed (changes requested):
This looks reasonable to me but I'd like some comments explaining why we set up `entrypoint.sh` like we do and documentation for the order in which this image tries to support CVMFS

DrDaveD commented on entrypoint.sh:
```diff
@@ -0,0 +1,25 @@
+    fi
+    $cvmfsexec_root/cvmfsexec -- /bin/true || \
+        fail "cvmfsexec smoke test failed.  You may not have the permissions to run cvmfsexec; see https://github.com/cvmfs/cvmfsexec#README for details"
+    exec $cvmfsexec_root/cvmfsexec $CVMFSEXEC_REPOS -- "$@"
```
There should also separately be some logic in the pilot, such that if it knows that a repository needs to be mounted, it isn't mounted, and $CVMFSMOUNT is not empty, then it should invoke $CVMFSMOUNT to mount it.  Also the pilot needs to invoke `exec {CVMFSEXEC_CMDFD}>&-` if $CVMFSEXEC_CMDFD is set, to prevent giving the job access to the cvmfsexec command.

DrDaveD replied:
@mambelli mmascher The GlideinWMS cvmfsexec support should check for cvmfsexec to already be running.   If it is, the pilot should skip invoking it, but still do the other things it needs to do to play well with cvmfsexec.

DrDaveD replied:
(@mmascher I accidentally left out the '@' for you on the previous comment.  I edited it in but I'm not sure if github would send a notice in response to an edit, so I add this extra comment.)

DrDaveD commented:
I think there also should be a CVMFS_HTTP_PROXY configuration variable available which, when set, allows the container user to specify a proxy or list of proxies to override the default in /cvmfsexec/dist/etc/cvmfs/default.local.  Also  CVMFS_QUOTA_LIMIT, if it is set, so people can easily give cvmfs more than the default 4GB of cache if they have the space.  Both of these options are called out in the [README](https://github.com/cvmfs/cvmfsexec#making-the-cvmfs-distribution).

matyasselmeci commented on entrypoint.sh:
```diff
@@ -0,0 +1,25 @@
+        echo "No cvmfsexec dir found for this user at $cvmfsexec_root; creating one"
+        cp -rp /cvmfsexec $cvmfsexec_root || fail "Couldn't create $cvmfsexec_root"
+        echo "Fetching $CVMFSEXEC_DIST CVMFS config"
+        $cvmfsexec_root/makedist $CVMFSEXEC_DIST || fail "Couldn't fetch CVMFS config"
```
I tried that, but the intent is for this image to be runnable as any UID, and I got problems when running as a different UID than the one that did the makedist.

DrDaveD replied:
If you do a makedist -o cvmfsexec (in a different directory or different name so as not overwrite cvmfsexec script) during the container build, then when cvmfsexec is executed it at run time it should self-extract as the user that is executing.  Alternatively maybe you could do a chown -R on the cvmfsexec directory.

DrDaveD replied:
On second thought I think using makedist -o isn't a good idea, because there ought to be a way for people to bind mount a new default.local, cache directory, and log directory.  So it's better to have it be unpacked.  However I don't understand why the cp -r to /tmp didn't make everything writable.

matyasselmeci replied:
makedist creates dist/var/lib/cvmfs with 0700 permissions so cp -r can't read it.

DrDaveD replied:
That comes from the cvmfs rpm so I guess you'll have to either chmod it, rmdir it and recreate it on the destination, or use a copy mechanism where you can exclude it.

Copying all the files to a new location complicates allowing people to bind mount in cache and log directories too.  I think it would probably be best to skip both dist/var/lib/cvmfs and dist/var/log and manipulate them manually.  If somebody has bind mounted either one at a documented location, create a symlink pointing back to them, otherwise I guess leave them in /tmp.  By looking at `mount` inside of a docker container it looks like the whole root filesystem is an overlay with the layers in /var/lib/docker/overlay2 so I guess that's OK as a default, but /var has to be big enough.  I think it would make sense to support bind-mounting at the paths /var/lib/cvmfs and /var/log/cvmfs, and also /etc/cvmfs/default.local.

matyasselmeci commented on entrypoint.sh:
```diff
@@ -0,0 +1,25 @@
+}
+
+shopt -s nullglob
+cvmfs_mounts=(/cvmfs/*)
```
Sure, sounds good.

matyasselmeci commented on entrypoint.sh:
```diff
@@ -0,0 +1,25 @@
+    cvmfsexec_root=/tmp/cvmfsexec-$(id -u)
+    if [[ ! -d $cvmfsexec_root ]]; then
+        echo "No cvmfsexec dir found for this user at $cvmfsexec_root; creating one"
+        cp -rp /cvmfsexec $cvmfsexec_root || fail "Couldn't create $cvmfsexec_root"
```
The copy in /tmp is owned by the user that's running the container, which may not be the user that created the container.

DrDaveD replied:
Maybe chown -R would be lighter weight.

DrDaveD replied:
Except that probably this script isn't running as fake root so chown wouldn't work.

matyasselmeci replied:
Exactly. I create a "template" in /cvmfsexec which each user (as themselves) copies to create their own cvmfsexec directory into which they have write permissions.

I wonder if it would be a good idea to run makedist at image creation, tar up the results and, at startup time have, have the user untar the results to create their own copy...

DrDaveD replied:
That's better than downloading at start time, but has the same issues as makedist -o I brought up in the [related thread on this ticket](https://github.com/opensciencegrid/osgvo-docker-pilot/pull/9#discussion_r599981246).  Follow up there.

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+If you don't have CVMFS available on the host, the container can still make
+CVMFS available by using [cvmfsexec](https://github.com/cvmfs/cvmfsexec#readme).
+
+This will require a kernel version >= 3.10.0-1127 on an EL7-compatible host
+or >= 4.18 on an EL8-compatible host, or user namespaces enabled by hand --
```
```suggestion
This will require a kernel version >= 3.10.0-1127 on an EL7-compatible host with user namespaces enabled
or >= 4.18 on an EL8-compatible host --
```

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+    `--security-opt seccomp=unconfined --security-opt systempaths=unconfined --device=/dev/fuse`
+    to the `docker run` invocation.
+
+The second option will add only the minimum necessary privileges for cvmfsexec.
```
```suggestion
The second option will add only the minimum necessary privileges for cvmfsexec.  You can add additional security to that option by also adding `--security-opt no-new-privileges`.
```

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+
+There are several environment variables you can set for cvmfsexec:
+
+-   `NO_CVMFSEXEC` - set this to disable cvmfsexec.
```
```suggestion
-   `NO_CVMFSEXEC` - set this to any non-empty value to disable cvmfsexec.
```

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+    add the above permissions to your container.  The default is to mount the
+    OASIS repo (oasis.opensciencegrid.org).
+
+-   `CVMFS_HTTP_PROXY` - this sets the proxy to use for CVMFS; leave this blank
+    to find the best one via wlcg-lpad.
```
```suggestion
-   `CVMFS_HTTP_PROXY` - this sets the proxy to use for CVMFS; if left blank
    it will find the best one via WLCG Web Proxy Auto Discovery.
```

DrDaveD commented on README.md:
````diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+
+```
+docker run -it --rm --user osg \
+       --cap-add=DAC_OVERRIDE --cap-add=SETUID --cap-add=SETGID \
+       --cap-add=CAP_DAC_READ_SEARCH \
+       --cap-add=SYS_ADMIN --cap-add=SYS_CHROOT --cap-add=SYS_PTRACE \
````
These capabilities are almost as bad as --privileged, especially SYS_ADMIN.  They should be unnecessary if you are using the options for low privilege.

matyasselmeci replied:
I copied these from the original example (which bind-mounted CVMFS). Perhaps @rynge would know why all of those are needed?

DrDaveD replied:
It was an attempt to do something less than `--privileged` but they didn't know about the completely unprivileged option.

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+
+
+If cvmfsexec does not have the required privileges, the container will fail.
+cvmfsexec will not be used if:
+- /cvmfs is already available via bind-mount
+- `CVMFSEXEC_REPOS` is empty
+- `NO_CVMFSEXEC` is set
```
Make clear this is an "or" list and not an "and" list.  Why have both NO_CVMFSEXEC and CVMFSEXEC_REPOS empty to disable?  Isn't NO_CVMFSEXEC redundant?

DrDaveD commented on README.md:
```diff
@@ -76,4 +76,80 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+to `/cvmfs-cache`.
+You can store the logs outside of the container by bind-mounting a directory to
+`/cvmfs-logs`.
+
```
```suggestion
You can customize cvmfs further by bind-mounting a file on
`/cvmfsexec/dist/etc/cvmfs/default.local`.
```

DrDaveD replied:
Oh but now I realize this won't work because of the tarball extraction.  You would have to choose another path.  How about `/cvmfsexec/default.local` and then look for that and copy it into the final path if present.

matyasselmeci replied:
Sure, that would work. If you bind-mount `/cvmfsexec/default.local` but also set the environment variables such as `CVMFS_HTTP_PROXY`, which should win?

DrDaveD replied:
I think you could do it whichever way is more convenient.  I would guess given your current implementation that the variables would win.

DrDaveD reviewed (changes requested):
I would just like to see support for bind mounting a user-defined default.local.  The rest looks good other than documentation.  I also suggest removing NO_CVMFSEXEC as unnecessary.

matyasselmeci commented:
NO_CVMFSEXEC is technically redundant, but as I was putting together the readme, I realized that disabling a feature as a side effect of emptying a list sounded confusing, so I wanted to make a single, obvious option for "I'm not ready for this yet, how do I turn it off so my container starts?"

This sounds a little weird, but I think another way to make it less confusing (instead of NO_CVMFSEXEC) would be to keep current behavior but make CVMFSEXEC_REPOS be empty by default. So the instructions to the admin would be like

> If you want CVMFS, either
> 1. Mount CVMFS on the host and pass `-v /cvmfs:/cvmfs` or
> 2. Pass `-e CVMFSEXEC_REPOS=oasis.opensciencegrid.org --security-opt ...`

Does that make sense? @brianhlin , do you have an opinion on this?

bbockelm commented:
Is there any obvious way to mount "all" CVMFS repositories?

DrDaveD commented:
> Is there any obvious way to mount "all" CVMFS repositories?

One could download the list of repositories hosted by oasis-replica.opensciencegrid.org (although it wouldn't be robust, what if it is down, what's the backup) or extract the list of supported OSG repositories from topology, but it would very likely be time-prohibitive.  cvmfsexec supports mounting additional repositories inside if the command it runs, so the better thing to do is for the pilot to request additional mounts when needed.

matyasselmeci commented:
OK, updated. I got rid of NO_CVMFSEXEC and instead made CVMFSEXEC_REPOS be blank by default. This way, an admin can just update their image without changing their `docker run` command to keep current behavior.

@DrDaveD , can you take another look?

matyasselmeci commented:
also @brianhlin, are my readme changes good?

DrDaveD commented:
Ok on the removal of NO_CVMFSEXEC but I'd still like to see support for /cvmfsexec/default.local and removal of all those extraneous --cap-add options.

brianhlin commented on README.md:
```diff
@@ -76,4 +83,78 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+with user namespaces enabled or >= 4.18 on an EL8-compatible host --
+see the cvmfsexec README linked above for details.
+
+This will also require granting the container some additional privileges, which
+you can do in one of two ways:
+
+1.  Add `--privileged` to the `docker run` invocation.
+
+2.  Add
+    `--security-opt seccomp=unconfined --security-opt systempaths=unconfined --device=/dev/fuse`
+    to the `docker run` invocation.
```
What's the use case for 1 vs 2?

DrDaveD replied:
The first case is if system admins are comfortable with the risk and don't want to enable unprivileged user namespaces.  The second case is less risky but requires enabling unprivileged user namespaces.

matyasselmeci replied:
2 uses fewer privileges but it might not be possible to do in Kubernetes (seccomp requires the cluster admin to set something up, and I found absolutely nothing on systempaths).

brianhlin replied:
@matyasselmeci please elaborate on 1 vs 2 using your + Dave's comments

brianhlin reviewed (commented):
Only one question about the README but otherwise it looks ok to me after a quick skim

DrDaveD commented:
You're not giving Kubernetes instructions at all, just docker, so I wouldn't mention that part unless you want to add an additional discussion on that.

Having said that, the equivalent of the systempaths=unconfined is apparently available in Kubernetes.  It is harder to enable, requiring setting a feature gate to enable [allowedProcMountTypes](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#allowedprocmounttypes).  I don't admin my own Kubernetes cluster so I haven't been able to try to get it to work myself.  One admin tried and failed, and Jeff said last week Tuesday he hoped to be ready to try it on Tiger by the time of the next kubernetes hackathon.

matyasselmeci commented:
OK, I added that `--privileged` was simpler and didn't require user namespaces, but gave more permissions to the container. I also explained what all the capabilities were used for (i.e. Singularity), and that `--privileged` implied them.

We can review which capabilities are needed and for what, but I would prefer to do that in a separate PR so we can keep the 'bind-mounted CVMFS' and 'CVMFS-via-cvmfsexec' command lines in sync.

I also added support for default.local.  Note that by default there is already stuff in `/cvmfsexec/dist/etc/cvmfs/default.local`:

```
CVMFS_HTTP_PROXY="auto;DIRECT"
CVMFS_PAC_URLS="http://grid-wpad/wpad.dat;http://wpad/wpad.dat;http://cernvm-wpad.fnal.gov/wpad.dat;http://cernvm-wpad.cern.ch/wpad.dat"
```
so anything the user mounts into `/cvmfsexec/default.local` will stomp over that.

DrDaveD commented on README.md:
```diff
@@ -16,14 +16,24 @@ In order to successfully start payload jobs:
+   This is the preferred mechanism.
+2. Use cvmfsexec as described in [the cvmfsexec section below](#cvmfs-without-a-bind-mount-using-cvmfsexec).
+
+Note: Supporting Singularity jobs inside the container will require the capabilities
```
No, this is still confused.  The same options needed for cvmfsexec work for singularity.  You either need to list the capabilities, use --privileged, or use the options for unprivileged.  I'm sorry I didn't make that clear; in fact I didn't realize that the discussion was distinct.

I suggest listing the 3 options here and not mention them under cvmfsexec.

matyasselmeci replied:
OK, I ended up skipping --privileged and just using the capabilities for both the bind-mount- and the cvmfsexec-based commands, and said that for cvmfsexec you need to add `--device=/dev/fuse`. I rebuilt the container to get cvmfsexec 4.8 and it works fine.

Do the changes look good?

DrDaveD commented:
> so anything the user mounts into `/cvmfsexec/default.local` will stomp over that.

Yes that's the way that cvmfsexec expects it.   It only supplies those defaults if default.local hasn't been supplied.

DrDaveD reviewed (changes requested):
Those things are good now but I thought of a couple of more suggestions.

One is that it would probably be good to include singularity.opensciencegrid.org also in the frequently used repos in addition to oasis.opensciencegrid.org.

The other is that you should probably close the CVMFSEXEC_CMDFD file descriptor as the cvmfsexec README advises, to prevent user payload jobs from attempting to mount more repositories.  cvmfsexec is designed to allow mounting more repos from within the entrypoint by executing $CVMFSMOUNT, but the idea was that the pilot wrapper would do it, and your pilot wrapper does not so it's probably too risky to leave it open.  I tried to find an easy way for you to do it and the easiest thing I found that would properly handle the quoting was to create a separate executable file wrapper.sh with one line
```
exec {CVMFSEXEC_CMDFD}>&- "$@"
```
and to change the `"$@"` on the last line of entrypoint.sh to be instead `wrapper.sh "$@"`.

matyasselmeci commented:
What if I edit `condor_master_wrapper` to add
```
if [[ -n $CVMFSEXEC_CMDFD ]]; then
    exec {CVMFSEXEC_CMDFD}>&-
fi
```
right before the `exec /usr/sbin/condor_master -f` line? Will that let me avoid making another wrapper?

DrDaveD commented:
Unfortunately that doesn't work because CVMFSEXEC_CMDFD is only set by cvmfsexec for the environment of the command it runs.  I experimented with a bunch of ways to try to get it into entrypoint.sh itself but everything messed up the quoting, for example if there's a blank in one of the parameters.  If you really don't want the extra file I could add an option to cvmfsexec to disable that feature.

Pilot scripts have a similar problem with wanting to execute some code inside singularity before the payload job.  The solutions I have seen used is to have singularity re-execute the same script that invokes singularity but behave differently based on an environment variable.  That's another option to creating a separate file, but more complicated.

matyasselmeci commented:
Yeah if it's not too hard please add the option to cvmfsexec.

DrDaveD commented:
Ok there's now a -N option in cvmfsexec-4.10.

matyasselmeci commented:
The -N option now makes cvmfsexec lock up after
```
CernVM-FS: loading Fuse module... done
CernVM-FS: mounted cvmfs on /cvmfsexec/dist/cvmfs/config-osg.opensciencegrid.org
```
Control-C'ing it gives me
```
^Ccvmfsexec smoke test failed.  You may not have the permissions to run cvmfsexec; see https://github.com/cvmfs/cvmfsexec#README for details
```

DrDaveD commented:
It's not clear to me how that could happen, and I can't reproduce it interactively.  What's the full command line you test with so I can hopefully reproduce it?

Anyway there isn't a need to use -N on the smoke test, but maybe you'll experience the same problem on the later invocation, I don't know.

matyasselmeci commented:
Yeah I added it to the smoke test because it failed on the later invocation as well.

Here's my command:
```
podman run --rm -it \
         --cap-add=DAC_OVERRIDE \
         --cap-add=DAC_READ_SEARCH \
         --cap-add=SETUID \
         --cap-add=SETGID \
         --cap-add=SYS_ADMIN \
         --cap-add=SYS_CHROOT \
         --cap-add=SYS_PTRACE \
         --device=/dev/fuse \
         -u 1000:1000 \
         -e CVMFSEXEC_REPOS=oasis.opensciencegrid.org \
          \
         matyasosg/osgvo-docker-pilot:cvmfsexec \
         bash
```
I just pushed a new image.

Looking at entrypoint.sh, what's being run is
```
tar -xzf /cvmfsexec.tar.gz -C /cvmfsexec
/cvmfsexec/cvmfsexec -N -- /bin/true
```

DrDaveD commented:
I ran that with docker on el7 and got different errors, resolved only by adding option `--security-opt seccomp=unconfined`.  I had noticed that earlier and mentioned it on slack but it should also get into the instructions here when adding cvmfsexec.  That is implied by `--privileged` and I thought it was implied by `--cap-add=SYS_ADMIN` as well but apparently it is not, at least on docker.  It's also needed for singularity so I think  it should be part of your singularity instructions as well.

However, once I added `--security-opt seccomp=unconfined` everything worked and I could not reproduce your problem.  It sounds like it is something specific to Fedora or podman, and I don't have a Fedora VM.  I have podman on my el7 VM, but it doesn't work properly and I get stuck on other problems.  Can you try it with docker on el7?  That's the primary platform we need to support.

DrDaveD commented:
I was able to reproduce the hang under podman on rocky8.

DrDaveD commented:
Alright this problem wasn't only in podman or docker, it was podman or docker without `--security-opts systempaths=unmasked`, which I was using when I tested with docker.  Fixed in cvmfsexec 4.11.

matyasselmeci commented:
Yep, can confirm that it's fixed.

matyasselmeci commented:
Anything else?

DrDaveD commented:
I think the cvmfsexec integration is good, although I think with docker it requires `--security-opt seccomp=unconfined`.   setuid-root singularity should be able to get away without it, but not cvmfsexec.  I cannot easily confirm at the moment because I don't think you have updated your matyasosg/osgvo-docker-pilot:cvmfsexec container on dockerhub with the latest cvmfsexec version.

Also I think singularity -cip will not work with the docker instructions you have, and that might hit you when the pilot tries to run a job under singularity.  In my testing I had to add `--security-opt systempaths=unconfined` with all those `--cap-adds` in order to unmask /proc, which singularity -p requires, I believe even when running setuid root.

matyasselmeci commented:
OK, I just updated the image.

Let's keep this PR focused on cvmfsexec and put any changes for singularity in a separate PR.

matyasselmeci commented:
I verified that I get errors for `docker` without `--security-opt=seccomp=unconfined` -- I see
```
pivot_root: failed to change root from `.' to `.old-root': Operation not permitted
mkdir: cannot create directory '/cvmfs': Permission denied
mount: mount point /cvmfs/config-osg.opensciencegrid.org does not exist
umount: /cvmfs/config-osg.opensciencegrid.org: mountpoint not found
rmdir: failed to remove '/cvmfs/config-osg.opensciencegrid.org': No such file or directory
```
so I updated the readme to mention that. Strange that it's not needed for podman.

DrDaveD commented:
Using your current matyasosg/osgvo-docker-pilot:cvmfsexec container if I try to run singularity I get an error 
```
ERROR  : No setuid installation found, for unprivileged installation use: ./mconfig --without-suid
```
Does it work for you?  This is under docker on centos7.  Using singularity -d indicates it's a problem with creating a socketpair:
```
DEBUG   [U=1000,P=593]     init()                        Create socketpair for master communication channel
DEBUG   [U=1000,P=593]     init()                        Create RPC socketpair for communication between stage 2 and RPC server
ERROR   [U=1000,P=593]     init()                        No setuid installation found, for unprivileged installation use: ./mconfig --without-suid
```

We sure are going through a lot of trouble to avoid --privileged.

matyasselmeci commented:
> Using your current matyasosg/osgvo-docker-pilot:cvmfsexec container if I try to run singularity I get an error
> 
> ```
> ERROR  : No setuid installation found, for unprivileged installation use: ./mconfig --without-suid
> ```
> 
> Does it work for you? This is under docker on centos7. Using singularity -d indicates it's a problem with creating a socketpair:
> 
> ```
> DEBUG   [U=1000,P=593]     init()                        Create socketpair for master communication channel
> DEBUG   [U=1000,P=593]     init()                        Create RPC socketpair for communication between stage 2 and RPC server
> ERROR   [U=1000,P=593]     init()                        No setuid installation found, for unprivileged installation use: ./mconfig --without-suid
> ```

Is this a new issue or does it also happen with, say `opensciencegrid/osgvo-docker-pilot:release`? If it's not new then I'd like to solve it in a separate PR.

> We sure are going through a lot of trouble to avoid --privileged.

I defer to @brianhlin and @rynge on why it was done that way, or if admins will accept just using --privileged and calling it a day.

DrDaveD commented on README.md:
```diff
@@ -17,12 +17,22 @@ In order to successfully start payload jobs:
+       --cap-add=DAC_READ_SEARCH \
        --cap-add=SYS_ADMIN --cap-add=SYS_CHROOT --cap-add=SYS_PTRACE \
        -v /cvmfs:/cvmfs:shared \
        -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org
```
```suggestion
       -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org \
```

DrDaveD commented on README.md:
```diff
@@ -77,4 +87,72 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+       --security-opt=seccomp=unconfined \
+       -v /var/cache/cvmfsexec:/cvmfsexec-cache \
+       -v /var/log/cvmfsexec:/cvmfsexec-logs \
+       -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org
```
```suggestion
       -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org \
```

DrDaveD commented:
Even with your container it doesn't fail if I pass in /cvmfs from the host.  It only fails following the cvmfsexec instructions.  I'm not sure why.  singularity exec -u for unprivileged works, except of course not with -p.  If I exchange all the --cap-adds and --security-opts for --privileged, singularity also works privileged.

matyasselmeci commented:
That's bizarre. OK, I'll use --privileged for the cvmfsexec invocation.

matyasselmeci commented:
@DrDaveD I changed the instructions for cvmfsexec to use `--privileged`. Does it look good now?

brianhlin commented on README.md:
```diff
@@ -17,15 +17,25 @@ In order to successfully start payload jobs:
+2. Use cvmfsexec as described in [the cvmfsexec section below](#cvmfs-without-a-bind-mount-using-cvmfsexec).
+
+Note: Supporting Singularity jobs inside the container will require the capabilities
+DAC_OVERRIDE, DAC_READ_SEARCH, SETGID, SETUID, SYS_ADMIN, SYS_CHROOT, and SYS_PTRACE.
```
```suggestion
`DAC_OVERRIDE`, `DAC_READ_SEARCH`, `SETGID`, `SETUID`, `SYS_ADMIN`, `SYS_CHROOT`, and `SYS_PTRACE`.
```

brianhlin commented on README.md:
```diff
@@ -77,4 +87,68 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+variable `CVMFSEXEC_REPOS`.
+
+On EL7, you must have kernel version >= 3.10.0-1127 (run `rpm -q kernel` to check),
+and user namespaces enabled.  On EL8, you must have kernel version >= 4.18.
```
For user namespaces, maybe we should point people to step 1 here https://opensciencegrid.org/docs/worker-node/install-singularity/#enabling-unprivileged-singularity?

brianhlin commented on README.md:
```diff
@@ -77,4 +87,68 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+
+On EL7, you must have kernel version >= 3.10.0-1127 (run `rpm -q kernel` to check),
+and user namespaces enabled.  On EL8, you must have kernel version >= 4.18.
+See the cvmfsexec README linked above for details.
```
```suggestion
See the [cvmfsexec README](https://github.com/cvmfs/cvmfsexec#readme) details.
```

May as well save people from having to hunt for the link

brianhlin commented on README.md:
```diff
@@ -77,4 +87,68 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+`/cvmfs-logs`.
+
+cvmfsexec requires the additional options `--device=/dev/fuse` and
+(for Docker) `--security-opt=seccomp=unconfined`.  Supporting Singularity
```
I don't understand what this parenthetical is referring to

brianhlin commented on README.md:
```diff
@@ -77,4 +87,68 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+Here is an example invocation using a token for authentication, using cvmfsexec
+to mount the OASIS and singularity repos instead of bind-mounting `/cvmfs`,
+sending the cache to `/var/cache/cvmfsexec` and the logs to
+`/var/log/cvmfsexec`, and adding support for Singularity jobs:
```
We should clarify that these paths are on the host

brianhlin reviewed (approved):
Some minor comments but otherwise LGTM!

DrDaveD commented on README.md:
```diff
@@ -77,4 +87,74 @@ $ singularity build osgvo-pilot.sif docker://opensciencegrid/osgvo-docker-pilot
+
+cvmfsexec requires the additional options `--device=/dev/fuse` and
+`--security-opt=seccomp=unconfined`.  Supporting Singularity jobs when using
+cvmfsexec requires privileged containers (`--privileged`).
```
```suggestion
cvmfsexec requires privileged containers (`--privileged`).  Using privileged containers avoids the need for those two additional options and all the `--cap-add` options without adding much risk so it is recommended.
```

DrDaveD reviewed (approved):
Approved after my one suggested addition to the documentation.
